app/models/rooms/open.rb

```diff
@@ -1,9 +1,18 @@
-# Rooms open to all users on the account. When a new user is added to the account, they're automatically granted membership.
-class Rooms::Open < Room
-  after_save_commit :grant_access_to_all_users
-
-  private
-    def grant_access_to_all_users
-      memberships.grant_to(User.active) if type_previously_changed?(to: "Rooms::Open")
-    end
-end
+# Rooms open to all users on the account. When a new user is added to the account, they're automatically granted membership.
+class Rooms::Open < Room
+  after_save_commit :grant_access_to_all_users
+
+  private
+    def grant_access_to_all_users
+      return unless type_previously_changed?(to: "Rooms::Open")
+
+      # Find the IDs of users who are already members of this room
+      existing_member_ids = self.memberships.pluck(:user_id)
+
+      # Find all active users who are NOT already members
+      users_to_add = User.active.where.not(id: existing_member_ids)

```

Comment:
Your current approach has a few performance drawbacks:
1. Uses two DB queries instead of one.
2. With a Campfire like Small Bets it results in creating large array objects in memory

This would be better as it also uses the existing index on `memberships(room_id, user_id, involvement)`:
```suggestion
      # Find active users who are NOT already members of this room
      users_to_add = User.active
                         .joins("LEFT JOIN memberships ON memberships.user_id = users.id AND memberships.room_id = #{id} AND memberships.active = true")
                         .where("memberships.id IS NULL")
```

Reply:
Thank you for the comment. You are right , i should have used a single query only. Issue addressed.